File: src/libjava/test/com/petroules/silverlock/DatabaseCryptoTest.java
package com.petroules.silverlock;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;
import java.util.Vector;

import javax.xml.parsers.ParserConfigurationException;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;
import org.spongycastle.crypto.CryptoException;
import org.spongycastle.crypto.PBEParametersGenerator;
import org.spongycastle.crypto.generators.PKCS5S2ParametersGenerator;
import org.spongycastle.crypto.params.KeyParameter;

import com.petroules.BufferedReader2;
import com.petroules.silverlock.DatabaseCrypto.CryptoStatus;
import com.petroules.silverlock.DatabaseCrypto.CryptoStatusWrapper;

public class DatabaseCryptoTest
{
	@BeforeClass
	public static void init()
	{
		TestData.setAssetManager(new FilesystemAssetManager("test-data"));
	}
	
	@Test
	public void testTwoWayFileEncryption() throws IOException, CryptoException
	{
		System.out.println(getTestAnnouncement("testTwoWayFileEncryption"));

		final String password = "MyS3cr3tP4$$w0rd";
		byte[] encryptedBytes = null;
		String encryptedString = null;

		StringWriter w = new StringWriter();
		BufferedWriter writer = new BufferedWriter(w);

		writeEncryptedData(writer, TestData.loremIpsum(), password);

		encryptedBytes = w.toString().getBytes("UTF-8");
		encryptedString = new String(encryptedBytes, "UTF-8");
		
		System.out.println(String.format("Encrypted file (%1$d bytes, %2$d characters): ", encryptedBytes.length, encryptedString.length()));
		System.out.println(encryptedString);

		byte[] encryptedBytesFromString = encryptedString.getBytes("UTF-8");
		Assert.assertArrayEquals("Byte array to string and back caused loss of data", encryptedBytes, encryptedBytesFromString);

		BufferedReader2 reader = new BufferedReader2(new StringReader(new String(encryptedBytesFromString, "UTF-8")));

		System.out.println("Decrypting...");
		String decryptedString = readEncryptedData(reader, password);

		System.out.println("Decrypted file:");
		System.out.println(decryptedString);

		Assert.assertEquals("Data did not decrypt back to original", TestData.loremIpsum(), decryptedString);
	}

	@Test
	public void testTwoWayEncryption() throws IOException, CryptoException
	{
		System.out.println(getTestAnnouncement("testTwoWayEncryption"));
		
		final String password = "MyS3cr3tP4$$w0rd";

		EncryptedData encryptedData = DatabaseCrypto.encrypt(password, TestData.loremIpsum());
		System.out.println("Encrypted data:");
		System.out.println("Salt: " + encryptedData.getSaltString());
		System.out.println("MAC: " + encryptedData.getMACString());
		System.out.println("Encrypted data: " + encryptedData.getDataString());

		String decryptedData = DatabaseCrypto.decrypt(password, encryptedData);
		System.out.println("Decrypted data: " + decryptedData);

		Assert.assertEquals("Data did not decrypt back to original", TestData.loremIpsum(), decryptedData);
	}

	@Test
	public void testTwoWayCompression() throws IOException
	{
		System.out.println(getTestAnnouncement("testTwoWayCompression"));
		
		byte[] bytes = TestData.loremIpsum().getBytes("UTF-8");

		try
		{
			byte[] compressed = DatabaseCrypto.compress(bytes);
			byte[] decompressed = DatabaseCrypto.decompress(compressed);
			Assert.assertArrayEquals("Data did not decompress back to original", bytes, decompressed);
		}
		catch (IOException e)
		{
			Assert.fail(e.getMessage());
		}
	}

	@Test
	public void testStandardFileDecryptsToExpectedXml()
	{
		System.out.println(getTestAnnouncement("testStandardFileDecryptsToExpectedXml"));
		
		Vector<TestDatabaseInfo> databases = TestData.testDatabases();
		for (TestDatabaseInfo item : databases)
		{
			DatabaseCrypto.CryptoStatusWrapper status = new CryptoStatusWrapper();
			String decryptedData = DatabaseCrypto.decrypt(item.databaseContents, item.password, status);
			System.out.println(decryptedData);
			Assert.assertEquals(CryptoStatus.NoError, status.cryptoStatus);
			Assert.assertEquals(item.xmlDatabaseContents, decryptedData);
		}
	}
	
	@Test
	public void testStandardFileDecryptsToInstance() throws IOException, ParserConfigurationException
	{
		System.out.println(getTestAnnouncement("testStandardFileDecryptsToInstance"));
		
		DatabaseReader reader = new DatabaseReader();
		
		Vector<TestDatabaseInfo> databases = TestData.testDatabases();
		for (TestDatabaseInfo item : databases)
		{
			Database db = reader.read(TestData.getAssetManager().open(item.fullPath), item.password);
			Assert.assertNotNull(reader.getErrorString(), db);
		}
	}
	
	@Test
	public void testHMAC()
	{
		System.out.println(getTestAnnouncement("testHMAC"));
		
		long start = System.currentTimeMillis();
		
		final PBEParametersGenerator generator = new PKCS5S2ParametersGenerator();
		generator.init(PBEParametersGenerator.PKCS5PasswordToBytes(("mypassword").toCharArray()), new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 }, 8192);
		byte[] key = ((KeyParameter)generator.generateDerivedParameters(256)).getKey();
		
		long time = System.currentTimeMillis() - start;
		Assert.assertTrue("Generated " + key.length + "-byte key, took " + time + "ms", true);
	}

	private String readEncryptedData(BufferedReader2 stream, String password) throws IOException, CryptoException
	{
		EncryptedData enc = new EncryptedData();
		stream.readLine(); // reads the header, we just ignore this
		enc.setSaltString(stream.readLine());
		enc.setMACString(stream.readLine());
		enc.setDataString(stream.readToEnd());

		return DatabaseCrypto.decrypt(password, enc);
	}

	private void writeEncryptedData(BufferedWriter stream, String data, String password) throws IOException, CryptoException
	{
		EncryptedData enc = DatabaseCrypto.encrypt(password, data);
		stream.write("SILVERLOCK DATABASE FILE 1.0 ZLIB");
		stream.newLine();
		stream.write(enc.getSaltString());
		stream.newLine();
		stream.write(enc.getMACString());
		stream.newLine();
		stream.write(enc.getDataString());
		stream.newLine();

		// VERY important... I wonder if this should be called by the caller
		// though?
		stream.flush();
	}
	
	private String getTestAnnouncement(String message)
	{
		final int lineLength = 80;
		
		StringBuilder builder = new StringBuilder();
		builder.append(getDashes(lineLength) + "\n");
		
		// The number of dashes to print after the length of
		// the message plus padding spaces are taken into account
		int dashCount = (lineLength - (message.length() + 2));
		
		builder.append(
				getDashes(dashCount / 2) +
				" " + message.toUpperCase() + " " +
				getDashes((dashCount % 2 == 0) ? (dashCount / 2) : (dashCount / 2 + 1)) +
				"\n");
		
		builder.append(getDashes(lineLength) + "\n");
		
		return builder.toString();
	}
	
	private String getDashes(int count)
	{
		StringBuilder builder = new StringBuilder();
		for (int i = 0; i < count; i++)
		{
			builder.append("-");
		}
		
		return builder.toString();
	}
}

File: src/lib/dotnet/DatabaseCryptoTest.cs
namespace Petroules.Silverlock.Test
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Globalization;
    using System.IO;
    using System.Linq;
    using System.Text;
    using Microsoft.VisualStudio.TestTools.UnitTesting;
    using Petroules.Silverlock.Test.Properties;

    [TestClass]
    public class DatabaseCryptoTest
    {
        [TestMethod]
        public void TestTwoWayFileEncryption()
        {
            const string password = "MyS3cr3tP4$$w0rd";
            byte[] encryptedBytes = null;
            string encryptedString = null;

            using (MemoryStream stream = new MemoryStream())
            using (StreamWriter writer = new StreamWriter(stream, Encoding.UTF8))
            {
                WriteEncryptedData(writer, TestData.LoremIpsum, password);

                encryptedBytes = stream.ToArray();
                encryptedString = Encoding.UTF8.GetString(encryptedBytes);
                Trace.WriteLine(string.Format(CultureInfo.InvariantCulture, "Encrypted file ({0} bytes, {1} characters): ", encryptedBytes.Length, encryptedString.Length));
                Trace.WriteLine(encryptedString);
            }

            byte[] encryptedBytesFromString = Encoding.UTF8.GetBytes(encryptedString);
            Assert.IsTrue(encryptedBytes.SequenceEqual(encryptedBytesFromString), "Byte array to string and back caused loss of data");

            using (MemoryStream stream = new MemoryStream(encryptedBytesFromString))
            using (StreamReader reader = new StreamReader(stream, Encoding.UTF8))
            {
                string decryptedString = ReadEncryptedData(reader, password);

                Trace.WriteLine("Decrypted file:");
                Trace.WriteLine(decryptedString);

                Assert.IsTrue(TestData.LoremIpsum == decryptedString, "Data did not decrypt back to original");
            }
        }

        [TestMethod]
        public void TestTwoWayEncryption()
        {
            const string password = "MyS3cr3tP4$$w0rd";

            EncryptedData encryptedData = DatabaseCrypto.Encrypt(password, TestData.LoremIpsum);
            Trace.WriteLine("Encrypted data:");
            Trace.WriteLine("Salt: " + encryptedData.SaltString);
            Trace.WriteLine("MAC: " + encryptedData.MACString);
            Trace.WriteLine("Encrypted data: " + encryptedData.DataString);

            string decryptedData = DatabaseCrypto.Decrypt(password, encryptedData);
            Trace.WriteLine("Decrypted data: " + decryptedData);

            Assert.IsTrue(TestData.LoremIpsum == decryptedData, "Data did not decrypt back to original");
        }

        [TestMethod]
        public void TestTwoWayCompression()
        {
            byte[] bytes = Encoding.UTF8.GetBytes(TestData.LoremIpsum);

            byte[] compressed = DatabaseCrypto.Compress(bytes);
            byte[] decompressed = DatabaseCrypto.Decompress(compressed);
            Assert.IsTrue(bytes.SequenceEqual(decompressed), "Data did not decompress back to original");
        }

        [TestMethod]
        public void TestStandardSilverlockDatabaseDecryption()
        {
            List<Tuple<string, string, string>> databases = new List<Tuple<string, string, string>>();
            databases.Add(new Tuple<string, string, string>(TestData.TestDatabasePassword, "test", TestData.TestDatabasePasswordXml));
            databases.Add(new Tuple<string, string, string>(TestData.TestDatabaseNoPassword, string.Empty, TestData.TestDatabaseNoPasswordXml));

            foreach (var item in databases)
            {
                /*using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes(item.Item1)))
                using (StreamReader reader = new StreamReader(stream, Encoding.UTF8))
                {
                    string decryptedData = ReadEncryptedData(reader, item.Item2);
                    Assert.AreEqual(item.Item3, decryptedData);
                    Trace.WriteLine(decryptedData);
                }*/

                DatabaseCrypto.CryptoStatus status;
                string decryptedData = DatabaseCrypto.Decrypt(item.Item1, item.Item2, out status);
                Trace.WriteLine(decryptedData);
                Assert.AreEqual(DatabaseCrypto.CryptoStatus.NoError, status);
                Assert.AreEqual(item.Item3, decryptedData);
            }
        }

        private string ReadEncryptedData(TextReader stream, string password)
        {
            EncryptedData enc = new EncryptedData();
            stream.ReadLine(); // reads the header, we just ignore this
            enc.SaltString = stream.ReadLine();
            enc.MACString = stream.ReadLine();
            enc.DataString = stream.ReadToEnd();

            return DatabaseCrypto.Decrypt(password, enc);
        }

        private void WriteEncryptedData(TextWriter stream, string data, string password)
        {
            EncryptedData enc = DatabaseCrypto.Encrypt(password, data);
            stream.WriteLine("SILVERLOCK DATABASE FILE 1.0 ZLIB");
            stream.WriteLine(enc.SaltString);
            stream.WriteLine(enc.MACString);
            stream.WriteLine(enc.DataString);

            // VERY important... I wonder if this should be called by the caller though?
            stream.Flush();
        }
    }
}
